File: src/com/googlecode/yadic/Container.scala
package com.googlecode.yadic

trait Container extends Resolver {
  def add[C](concrete: Class[C]): Container

  def add[I, C <: I](interface: Class[I], concrete: Class[C]): Container

  def add[T](a: Class[T], activator: Activator[T]): Container

  def addInstance(instance: Object): Container
  
  def add[T](aClass: Class[T], activator: () => T): Container

  def decorate[I, C <: I](interface: Class[I], concrete: Class[C]): Container

  def remove[T](aClass: Class[T]): Activator[T]

  def contains[T](aClass: Class[T]):Boolean

  def resolveType[A](aClass: Class[A]): A
}

File: src/com/googlecode/yadic/Container.java
package com.googlecode.yadic;



import com.googlecode.totallylazy.Function1;

import java.util.concurrent.Callable;

public interface Container extends TypeMap {
    <C> Container add(Class<C> concrete);

    <I, C extends I> Container add(Class<I> anInterface, Class<C> concrete);

    <I, C extends I> Container addInstance(Class<I> anInterface, C instance);

    <T, A extends Callable<T>> Container addActivator(Class<T> aClass, Class<A> activator);

    <T> Container addActivator(Class<T> aClass, Callable<? extends T> activator);

    <I, C extends I> Container decorate(Class<I> anInterface, Class<C> concrete);

    <T> T get(Class<T> aClass);

    <T> Callable<T> getActivator(Class<T> aClass);

    <I, C extends I> Container replace(Class<I> anInterface, Class<C> newConcrete);

    class functions{
        public static <T> Function1<Container, T> get(final Class<T> aClass) {
            return container -> container.get(aClass);
        }
    }
}